p2p/pex/pex_reactor.go

```diff
@@ -585,81 +592,65 @@ func (r *PEXReactor) crawlPeersRoutine() {
 func (r *PEXReactor) hasPotentialPeers() bool {
 	out, in, dial := r.Switch.NumPeers()
 
-	return out+in+dial > 0 && len(r.book.ListOfKnownAddresses()) > 0
+	return out+in+dial > 0 && r.book.Size() > 0
 }
 
-// crawlPeerInfo handles temporary data needed for the
-// network crawling performed during seed/crawler mode.
+// crawlPeerInfo handles temporary data needed for the network crawling
+// performed during seed/crawler mode.
 type crawlPeerInfo struct {
-	// The listening address of a potential peer we learned about
 	Addr *p2p.NetAddress
-
-	// The last time we attempt to reach this address
-	LastAttempt time.Time
-
-	// The last time we successfully reached this address
-	LastSuccess time.Time
+	// The last time we crawled the peer or attempted to do so.
+	LastCrawled time.Time
 }
 
-// oldestFirst implements sort.Interface for []crawlPeerInfo
-// based on the LastAttempt field.
-type oldestFirst []crawlPeerInfo
-
-func (of oldestFirst) Len() int           { return len(of) }
-func (of oldestFirst) Swap(i, j int)      { of[i], of[j] = of[j], of[i] }
-func (of oldestFirst) Less(i, j int) bool { return of[i].LastAttempt.Before(of[j].LastAttempt) }
-
-// getPeersToCrawl returns addresses of potential peers that we wish to validate.
-// NOTE: The status information is ordered as described above.
-func (r *PEXReactor) getPeersToCrawl() []crawlPeerInfo {
-	var of oldestFirst
+// CrawlPeers will crawl the network looking for new peer addresses.
+// Exposed for testing.
+func (r *PEXReactor) CrawlPeers(addrs []*p2p.NetAddress) {
+	now := time.Now()
 
-	// TODO: be more selective
-	addrs := r.book.ListOfKnownAddresses()
 	for _, addr := range addrs {
-		if len(addr.ID()) == 0 {
-			continue // dont use peers without id
+		peerInfo, ok := r.crawlPeerInfos[addr.ID]
+		if !ok {
+			peerInfo = &crawlPeerInfo{
+				Addr:        addr,
+				LastCrawled: now.Add(-defaultCrawlPeerInterval).Add(-1 * time.Second),
+			}
+			r.crawlPeerInfos[addr.ID] = peerInfo
 		}
 
-		of = append(of, crawlPeerInfo{
-			Addr:        addr.Addr,
-			LastAttempt: addr.LastAttempt,
-			LastSuccess: addr.LastSuccess,
-		})
-	}
-	sort.Sort(of)
-	return of
-}
+		// Do not attempt to connect with peers we recently crawled.
+		if now.Sub(peerInfo.LastCrawled) < defaultCrawlPeerInterval {
+			continue
+		}
 
-// crawlPeers will crawl the network looking for new peer addresses. (once)
-func (r *PEXReactor) crawlPeers() {
-	peerInfos := r.getPeersToCrawl()
+		// Record crawling attempt.
+		peerInfo.LastCrawled = now
+		// XXX: - grows linearly with a number of peers crawled
+		//      - does not get persisted
```

Comment:
Sounds like this might eventually cause trouble. Maybe open an issue for this too? (I do not know how problematic this might become though)